storage: fix TestStoreRangeCorruptionChangeReplicas

BramGruneir opened the pull request:
2 commits:
1. fixes storepool.deadReplicas()
2. ensures that storeList won't assign a new replica to replace a currently dead(corrupt) one on a store

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/10141)

<!-- Reviewable:end -->

BramGruneir committed:
storage: fix deadReplicas()

After the storeReplicaVisitor was updated to not hold the store lock the whole
time, it also no longer returns dead replicas. This fixes the deadReplicas()
function.

This also fixes the first type of error found in #8664:
`expected len(replicas) = 3, got [{NodeID:1 StoreID:1 ReplicaID:1} {NodeID:6 StoreID:6 ReplicaID:2}]`

Part of #8664.

BramGruneir committed:
storage: get the store pool to correctly deal with dead replicas

When a replica is corrupt on a store, we will no longer try to allocate a new
replica on it for the same range as the dead replica. I've also added some unit
tests that also ensure that the allocator and store pool act correctly.

This fixes the 2nd error in #8664:
`expected new replica to not be placed back on same store as corrupt replica`

Fixes #8664

tamird commented:
Well, certainly seems like strong evidence that fixing flaky tests is important.

---

Reviewed 2 of 2 files at r1, 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 15 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator_test.go, line 237 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtnpMUNMjKFrtLzvl:-KUdtnpNcKbV3U_ZWbi2:b-o1156s) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L237)):*
> ```Go
> 	gossiputil.NewStoreGossiper(g).GossipStores(sameDCStores, t)
> 
> 	// Set the Store1 to have a dead store in the store pool.
> ```

s/the Store1/a store/

---

*[pkg/storage/allocator_test.go, line 250 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtxZm4CrJbD3Pks6h:-KUdtxZm4CrJbD3Pks6i:bx987xx) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L250)):*
> ```Go
> 	)
> 	if err != nil {
> 		t.Fatalf("Unable to perform allocation: %v", err)
> ```

just `t.Fatal(err)` seems fine. `%v` implies that `err` may be nil.

---

*[pkg/storage/allocator_test.go, line 253 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdu5foTUOucXQ3OOyF:-KUdu5fpnOYjaLxcAsG8:b-2ie09q) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L253)):*
> ```Go
> 	}
> 	if result.Node.NodeID != 2 || result.StoreID != 2 {
> 		t.Errorf("expected NodeID 2 and StoreID 2: %+v", result)
> ```

is this really a reasonable test? in other words, isn't it possible that this will be the outcome even if the allocator logic is incorrect, in the case where the allocator is even slightly non-deterministic?

---

*[pkg/storage/client_raft_test.go, line 836 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdrQPlHhLbmNQHQr95:-KUdrQPmagKcdWItMGnB:b-2074ia) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6f437bb6feb6d76755972d5880adfdb13fcbfe4/pkg/storage/client_raft_test.go#L836)):*
> ```Go
> 	defer mtc.Stop()
> 	mtc.initGossipNetwork()
> ```

what's with these random whitespace changes?

---

*[pkg/storage/client_raft_test.go, line 875 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdrSJTWd8GI-pSs-9d:-KUdrSJUyKrOijgShLMQ:b-2074ia) ([raw file](https://github.com/cockroachdb/cockroach/blob/56027bad258fb830b2759dbd5ac6ba800b48f231/pkg/storage/client_raft_test.go#L875)):*
> ```Go
> 			t.Fatal(err)
> 		}
> ```

what's with these random whitespace changes?

---

*[pkg/storage/replicate_queue.go, line 126 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdsqy2rZrtrvS-D6T4:-KUdsqy2rZrtrvS-D6T5:b-8gou01) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/replicate_queue.go#L126)):*
> ```Go
> 		desc.Replicas,
> 		leaseStoreID,
> 		repl.RangeID,
> ```

desc.RangeID?

---

*[pkg/storage/replicate_queue.go, line 164 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdspsq6F44Hdft5985:-KUdspsq6F44Hdft5986:b-8gou01) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/replicate_queue.go#L164)):*
> ```Go
> 			zone.Constraints,
> 			desc.Replicas,
> 			repl.RangeID,
> ```

desc.RangeID?

---

*[pkg/storage/replicate_queue.go, line 219 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdss4iTr5mfjKwf5y0:-KUdss4iTr5mfjKwf5y1:b-8gou01) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/replicate_queue.go#L219)):*
> ```Go
> 			desc.Replicas,
> 			repl.store.StoreID(),
> 			repl.RangeID,
> ```

desc.RangeID?

---

*[pkg/storage/store.go, line 2024 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdrbZ-djhAATQtg_kQ:-KUdrbZ0PfzjrclJP_o7:bysyhus) ([raw file](https://github.com/cockroachdb/cockroach/blob/56027bad258fb830b2759dbd5ac6ba800b48f231/pkg/storage/store.go#L2024)):*
> ```Go
> 	// processing.
> 	s.mu.Lock()
> 	replicas := make([]*Replica, 0, len(s.mu.replicas))
> ```

nit: `append` is slightly slower than:

```
replicas := make([]*Replica, len(s.mu.replicas))
for i, repl := range s.mu.replicas {
	replicas[i] = repl
}
```

---

*[pkg/storage/store.go, line 2036 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdrznxLhrunC4w5DAd:-KUdrzny9b6zH6S1k4n0:b-eid7ae) ([raw file](https://github.com/cockroachdb/cockroach/blob/56027bad258fb830b2759dbd5ac6ba800b48f231/pkg/storage/store.go#L2036)):*
> ```Go
> 		desc := r.mu.state.Desc
> 		r.mu.Unlock()
> 		replicaDesc, ok := desc.GetReplicaDescriptor(s.Ident.StoreID)
> ```

this change is immaterial, isn't it? Seems like random noise to have it here.

---

*[pkg/storage/store.go, line 2045 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdsZTdcOV_AIAB2KGu:-KUdsZTdcOV_AIAB2KGv:bq2wt3t) ([raw file](https://github.com/cockroachdb/cockroach/blob/56027bad258fb830b2759dbd5ac6ba800b48f231/pkg/storage/store.go#L2045)):*
> ```Go
> 	}
> 	return roachpb.StoreDeadReplicas{
> 		StoreID:  s.Ident.StoreID,
> ```

ditto, this should probably use the same `sid`

---

*[pkg/storage/store_pool.go, line 99 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdt5DIqbDbchn3Sa4X:-KUdt5DIqbDbchn3Sa4Y:b-kqh69l) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool.go#L99)):*
> ```Go
> // These are the possible values for a storeMatch.
> const (
> 	// storeMatchDead - the store is not yet available or has been timed out.
> ```

there is no need to repeat these names, none of these are exported anyway.

this also seems like noise in this change; it's hard to spot the new variant =/

---

*[pkg/storage/store_pool_test.go, line 326 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtSUDxr1j67odoQX8:-KUdtSUDxr1j67odoQX9:b3ajvbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool_test.go#L326)):*
> ```Go
> 
> 	sp.mu.Lock()
> 	// Set the deadStore as dead.
> ```

either "the dead store" or "deadStore"

---

*[pkg/storage/store_pool_test.go, line 328 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtWICQwzljP_unz-W:-KUdtWICQwzljP_unz-X:b-7ropln) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool_test.go#L328)):*
> ```Go
> 	// Set the deadStore as dead.
> 	sp.mu.storeDetails[deadStore.StoreID].markDead(sp.clock.Now())
> 	// Set the declinedStore as throttled.
> ```

ditto

---

*[pkg/storage/store_pool_test.go, line 330 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtX6mAbnO6jfGc_bz:-KUdtX6mAbnO6jfGc_c-:bqlswgc) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool_test.go#L330)):*
> ```Go
> 	// Set the declinedStore as throttled.
> 	sp.mu.storeDetails[declinedStore.StoreID].throttledUntil = sp.clock.Now().GoTime().Add(time.Hour)
> 	// Add a corrupt replica to the corruptReplicaStore.
> ```

again, be consistent, if you're using "deadStore" above, use "corruptReplica" here rather than "a corrupt replica". That said, I think it's better to avoid quoting variable names directly in all these cases to reduce probability of rot.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-:-KUduHlqe0wtyvmGlFWV:b-5lo53s)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Reviewed 2 of 2 files at r1, 8 of 8 files at r2.
Review status: all files reviewed at latest revision, 19 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator.go, line 207 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUluWu3Bh3J6tN3wSOt:-KUluWu3Bh3J6tN3wSOu:b-n21kc7) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator.go#L207)):*
> ```Go
> 	constraints config.Constraints,
> 	existing []roachpb.ReplicaDescriptor,
> 	rangeID roachpb.RangeID,
> ```

These range ID arguments should be documented, and in particular the fact that 0 is a legal value (which I found surprising)

---

*[pkg/storage/allocator_test.go, line 237 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUlvx37zQUJlKKd2MAA:-KUlvx37zQUJlKKd2MAB:b-kpp68m) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L237)):*
> ```Go
> 	gossiputil.NewStoreGossiper(g).GossipStores(sameDCStores, t)
> 
> 	// Set the Store1 to have a dead store in the store pool.
> ```

s/dead store/dead replica/

---

*[pkg/storage/allocator_test.go, line 240 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUlv3L8whCDJB9BViZv:-KUlv3L9NXZrdbayBrRT:b-u7r17j) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L240)):*
> ```Go
> 	sp.mu.Lock()
> 	sp.mu.storeDetails[roachpb.StoreID(1)].deadReplicas[roachpb.RangeID(1)] =
> 		[]roachpb.ReplicaDescriptor{{}}
> ```

What does the empty replica descriptor signify here? It's strange that we don't at least fill in the store ID with the known value. We should generally try to fill in all the fields that will be filled in in production. 

---

*[pkg/storage/store_pool_test.go, line 227 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUlwqrbaJqjlCGP0IX5:-KUlwqrbaJqjlCGP0IX6:b-kr98sf) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool_test.go#L227)):*
> ```Go
> 	sl, aliveStoreCount, throttledStoreCount := sp.getStoreList(
> 		constraints,
> 		roachpb.RangeID(1), // To match the corrupt store in TestSTorePoolGetSToreList.
> ```

This is too subtle. TestStorePoolGetStoreList uses both RangeID(0) and RangeID(1) and it's not obvious which uses need to match with this one. verifyStoreList should take a range ID argument and then TestStorePoolGetStoreList can define constants for 0 and 1 and clarify what's going on. There should probably also be some verification that a replica that has not been marked corrupt is unaffected (or that two ranges can have corrupt replicas on different stores). 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Indeed.  In this case, it went from just being a flaky test to a full on bug due to a change that occurred after the test was disabled.

---

Review status: 0 of 10 files reviewed at latest revision, 19 unresolved discussions.

---

*[pkg/storage/allocator.go, line 207 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUluWu3Bh3J6tN3wSOt:-KUrEF6RLdFf5p_wUHwA:b-2w104i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator.go#L207)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> These range ID arguments should be documented, and in particular the fact that 0 is a legal value (which I found surprising)
</details>
Done.  RangeID of 0 is only used in tests to avoid checking if there's a dead replica.  It's not a valid value.  I've changed them all and fixed a few of the storeIDs of 0 and -1 as well.

---

*[pkg/storage/allocator_test.go, line 237 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtnpMUNMjKFrtLzvl:-KUrFhp5J8kU5-gY-nyo:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L237)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> s/the Store1/a store/
</details>
Done.

---

*[pkg/storage/allocator_test.go, line 237 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUlvx37zQUJlKKd2MAA:-KUrFipemNAQycxy_RFi:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L237)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> s/dead store/dead replica/
</details>
Done.

---

*[pkg/storage/allocator_test.go, line 240 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUlv3L8whCDJB9BViZv:-KUrFni55QMIq5q6TEav:bwxfkk4) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L240)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> What does the empty replica descriptor signify here? It's strange that we don't at least fill in the store ID with the known value. We should generally try to fill in all the fields that will be filled in in production. 
</details>
The check only looks to see if any dead replicas exist, and it's already a map keyed by storeID, then by rangeID.  So the details are not relevant.  However, I've added in the details anyway, in case we add new logic to the dead replica check.

---

*[pkg/storage/allocator_test.go, line 250 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtxZm4CrJbD3Pks6h:-KUrHVgvCCxywlfrgynv:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L250)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> just `t.Fatal(err)` seems fine. `%v` implies that `err` may be nil.
</details>
Done.

---

*[pkg/storage/allocator_test.go, line 253 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdu5foTUOucXQ3OOyF:-KUrJirYBBnWkjmFF9yb:b-hdjbyk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L253)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> is this really a reasonable test? in other words, isn't it possible that this will be the outcome even if the allocator logic is incorrect, in the case where the allocator is even slightly non-deterministic?
</details>
It is deterministic.  The flag is set to true. 

---

*[pkg/storage/client_raft_test.go, line 836 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdrQPlHhLbmNQHQr95:-KUrJvGizJqMnStN4-Ev:b-iiig0i) ([raw file](https://github.com/cockroachdb/cockroach/blob/e6f437bb6feb6d76755972d5880adfdb13fcbfe4/pkg/storage/client_raft_test.go#L836)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's with these random whitespace changes?
</details>
I had peppered the test with outputs, then removed them.  That's probably the cause.
This one can be removed.  So I'll leave it as such.

---

*[pkg/storage/client_raft_test.go, line 875 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdrSJTWd8GI-pSs-9d:-KUrK7WrM9S-10hVxWWd:b1o60wu) ([raw file](https://github.com/cockroachdb/cockroach/blob/56027bad258fb830b2759dbd5ac6ba800b48f231/pkg/storage/client_raft_test.go#L875)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> what's with these random whitespace changes?
</details>
This reads a lot better with the space before the comment.  

---

*[pkg/storage/replicate_queue.go, line 126 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdsqy2rZrtrvS-D6T4:-KUrgMx69148T2SdqUOf:b-mz131d) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/replicate_queue.go#L126)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> desc.RangeID?
</details>
sure, they should be the same.

---

*[pkg/storage/replicate_queue.go, line 164 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdspsq6F44Hdft5985:-KUrgQsfiy0kfHyZZx7O:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/replicate_queue.go#L164)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> desc.RangeID?
</details>
Done.

---

*[pkg/storage/replicate_queue.go, line 219 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdss4iTr5mfjKwf5y0:-KUrgZlEhB_wkiPQujhp:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/replicate_queue.go#L219)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> desc.RangeID?
</details>
Done.

---

*[pkg/storage/store.go, line 2024 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdrbZ-djhAATQtg_kQ:-KUrhJNGV-q2l1lncg1q:b-epbvnc) ([raw file](https://github.com/cockroachdb/cockroach/blob/56027bad258fb830b2759dbd5ac6ba800b48f231/pkg/storage/store.go#L2024)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: `append` is slightly slower than:
> 
> ```
> replicas := make([]*Replica, len(s.mu.replicas))
> for i, repl := range s.mu.replicas {
> 	replicas[i] = repl
> }
> ```
</details>
Can't, replicas is a map, not a slice.  Perhaps one could make the argument that making a copy of the map would be faster, I'm not sure on that one.

---

*[pkg/storage/store.go, line 2036 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdrznxLhrunC4w5DAd:-KUrh_8HdquQD-zh_gzU:b-74dkx2) ([raw file](https://github.com/cockroachdb/cockroach/blob/56027bad258fb830b2759dbd5ac6ba800b48f231/pkg/storage/store.go#L2036)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> this change is immaterial, isn't it? Seems like random noise to have it here.
</details>
I removed an unneeded local variable.  How is that immaterial? 
Do you want it as two different commits?  I'm confused about what you're looking for here.

---

*[pkg/storage/store.go, line 2045 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdsZTdcOV_AIAB2KGu:-KUrhmS-mJt0YSGSrqHl:b-1zidty) ([raw file](https://github.com/cockroachdb/cockroach/blob/56027bad258fb830b2759dbd5ac6ba800b48f231/pkg/storage/store.go#L2045)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto, this should probably use the same `sid`
</details>
again, not sure why this is a big deal.  It should have used sid originally, and didn't.

---

*[pkg/storage/store_pool.go, line 99 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdt5DIqbDbchn3Sa4X:-KUri4TYcJyxXxs7sW3V:bfg3ewh) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool.go#L99)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> there is no need to repeat these names, none of these are exported anyway.
> 
> this also seems like noise in this change; it's hard to spot the new variant =/
</details>
I tried to clean it up.  I'll remove the repeated names.  That should make your review easier.

---

*[pkg/storage/store_pool_test.go, line 227 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUlwqrbaJqjlCGP0IX5:-KUrljE6U_GrBv2sof_x:b-cs1084) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool_test.go#L227)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary>
> This is too subtle. TestStorePoolGetStoreList uses both RangeID(0) and RangeID(1) and it's not obvious which uses need to match with this one. verifyStoreList should take a range ID argument and then TestStorePoolGetStoreList can define constants for 0 and 1 and clarify what's going on. There should probably also be some verification that a replica that has not been marked corrupt is unaffected (or that two ranges can have corrupt replicas on different stores). 
</details>
you're right, I've hopefully made it a lot cleaner, added the negative case at the start.  This should have tested both the negative and positive case.  I've also added a bunch of unrelated corrupted replicas to other stores.

---

*[pkg/storage/store_pool_test.go, line 326 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtSUDxr1j67odoQX8:-KUrjh_v7ab7Na8EF_KZ:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool_test.go#L326)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> either "the dead store" or "deadStore"
</details>
Done.

---

*[pkg/storage/store_pool_test.go, line 328 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtWICQwzljP_unz-W:-KUrji0_a1JFUNTPm0nX:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool_test.go#L328)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> ditto
</details>
Done.

---

*[pkg/storage/store_pool_test.go, line 330 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtX6mAbnO6jfGc_bz:-KUrjqIkNNPO9_mj51cR:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool_test.go#L330)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> again, be consistent, if you're using "deadStore" above, use "corruptReplica" here rather than "a corrupt replica". That said, I think it's better to avoid quoting variable names directly in all these cases to reduce probability of rot.
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-:-KUrkXj0mM6Ut2bAahc4:b-ndxy1l)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 10 files at r3, 8 of 8 files at r4.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks pending.

---

*[pkg/storage/allocator_test.go, line 237 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtnpMUNMjKFrtLzvl:-KUsScZiwAZxyS51LtyD:bstxrrv) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L237)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Done.
</details>
You seem to have done something different than my suggestion. The point was to something that looks like a variable name since that is susceptible to rot. "the Store" is odd for a different reason: there is more than one store in the context of this test (and we don't usually capitalize "store").

---

*[pkg/storage/allocator_test.go, line 240 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUlv3L8whCDJB9BViZv:-KUsT4hOhWiiEHa06_-5:bs7zxzk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L240)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> The check only looks to see if any dead replicas exist, and it's already a map keyed by storeID, then by rangeID.  So the details are not relevant.  However, I've added in the details anyway, in case we add new logic to the dead replica check.
</details>
Can you extract some constants? An awful lot of instances of `roachpb.{Node,Store,Replica}ID(1)` are now present here.

---

*[pkg/storage/replicate_queue.go, line 126 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdsqy2rZrtrvS-D6T4:-KUsTZQJ12P2mlgvrf46:b-svq80i) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/replicate_queue.go#L126)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> sure, they should be the same.
</details>
OK, but this is still `repl.RangeID`, which is inconsistent with all the surrounding code.

---

*[pkg/storage/store.go, line 2024 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdrbZ-djhAATQtg_kQ:-KUsThLWNpKyMgBSNeZP:b-q0sif) ([raw file](https://github.com/cockroachdb/cockroach/blob/56027bad258fb830b2759dbd5ac6ba800b48f231/pkg/storage/store.go#L2024)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Can't, replicas is a map, not a slice.  Perhaps one could make the argument that making a copy of the map would be faster, I'm not sure on that one.
</details>
Ah, didn't realize. Copying the map is almost certainly slower.

---

*[pkg/storage/store.go, line 2036 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdrznxLhrunC4w5DAd:-KUsTpvwKJ_6TYQeXNyS:b-d5302) ([raw file](https://github.com/cockroachdb/cockroach/blob/56027bad258fb830b2759dbd5ac6ba800b48f231/pkg/storage/store.go#L2036)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> I removed an unneeded local variable.  How is that immaterial? 
> Do you want it as two different commits?  I'm confused about what you're looking for here.
</details>
I'm telling you that changes like this make it more difficult to review your changes. Removing a local variable is definitely immaterial to the other changes here. Indeed, I'd have put it in a separate commit for this reason.

---

*[pkg/storage/store_pool_test.go, line 330 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtX6mAbnO6jfGc_bz:-KUsUxwZZXRQlfLAPd6Z:bmk0xwk) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool_test.go#L330)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Done.
</details>
Heh, so you went with the more rot-prone option.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 12 unresolved discussions, all commit checks successful.

---

*[pkg/storage/allocator_test.go, line 237 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtnpMUNMjKFrtLzvl:-KUs_CemPofFZGcJ25oh:bvj04xs) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L237)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> You seem to have done something different than my suggestion. The point was to something that looks like a variable name since that is susceptible to rot. "the Store" is odd for a different reason: there is more than one store in the context of this test (and we don't usually capitalize "store").
</details>
Bah, you're right.  I thought I fixed this, but I think I reverted it while looking for the causes of some test errors.

It's not just any store, but specifically store 1 that gets the dead replica.  

---

*[pkg/storage/allocator_test.go, line 240 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUlv3L8whCDJB9BViZv:-KUtImz9MK8621xN5TML:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/allocator_test.go#L240)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Can you extract some constants? An awful lot of instances of `roachpb.{Node,Store,Replica}ID(1)` are now present here.
</details>
Done.

---

*[pkg/storage/replicate_queue.go, line 126 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdsqy2rZrtrvS-D6T4:-KUtJ11YcTFj8bsQn7gg:b-o8v1xt) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/replicate_queue.go#L126)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> OK, but this is still `repl.RangeID`, which is inconsistent with all the surrounding code.
</details>
I thought I changed this one.  Again, probably an error in revert to track down where a new test failure was coming from.  Fixed

---

*[pkg/storage/store.go, line 2036 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdrznxLhrunC4w5DAd:-KUtJDBNwCzyzQrVkYQn:b-fysmdi) ([raw file](https://github.com/cockroachdb/cockroach/blob/56027bad258fb830b2759dbd5ac6ba800b48f231/pkg/storage/store.go#L2036)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> I'm telling you that changes like this make it more difficult to review your changes. Removing a local variable is definitely immaterial to the other changes here. Indeed, I'd have put it in a separate commit for this reason.
</details>
OK, next time I will.  I'm going to leave it as is since this is already done.

---

*[pkg/storage/store_pool_test.go, line 330 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUdtX6mAbnO6jfGc_bz:-KUtJPJugjWBucGuhEH5:brt3r2v) ([raw file](https://github.com/cockroachdb/cockroach/blob/a072d8dfaf6762276c9679f4eb96d62e0e851290/pkg/storage/store_pool_test.go#L330)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> Heh, so you went with the more rot-prone option.
</details>
Indeed.  In this case it makes a lot more sense since these cases are directly related to the variables above.  They're not used outside the function and the chance of rot is minimal.
I tried it the other way and it just didn't convey why the changes were being made.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r5.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/allocator_test.go, line 239 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUtZiGojTOTcies14qf:-KUtZiGpdg1V9AfT57fC:b-shjyfj) ([raw file](https://github.com/cockroachdb/cockroach/blob/702e047e43132ce8d38a839c15196aeb9ebaa64f/pkg/storage/allocator_test.go#L239)):*
> ```Go
> 	defer stopper.Stop()
> 	gossiputil.NewStoreGossiper(g).GossipStores(sameDCStores, t)
> 	store1ID := roachpb.StoreID(1)
> ```

nit: can be a const

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[pkg/storage/allocator_test.go, line 239 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-KUtZiGojTOTcies14qf:-KUteWYA5hBq2IhAbakb:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/702e047e43132ce8d38a839c15196aeb9ebaa64f/pkg/storage/allocator_test.go#L239)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> nit: can be a const
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
So am I good to merge this?

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-:-KUxPy5z7HSLdZoyEw7b:b-56i0xh)*
<!-- Sent from Reviewable.io -->

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/10141#-:-KUxQx4D-DvcsbX69LL-:bnfp4nl)*
<!-- Sent from Reviewable.io -->
